Implement navigation rate limiting

roberto-apple opened the pull request:
#### 923205935b5be77b722f998a306732eacd648754
<pre>
Implement navigation rate limiting
<a href="https://bugs.webkit.org/show_bug.cgi?id=302355">https://bugs.webkit.org/show_bug.cgi?id=302355</a>
<a href="https://rdar.apple.com/164510890">rdar://164510890</a>

Reviewed by Basuke Suzuki.

Add a sliding window rate limiter (200 navigations per 10 second window) to the Navigation API
to prevent malicious or buggy scripts from flooding the system with navigation requests. This
matches Chromium&apos;s implementation and prevents IPC flooding, resource exhaustion, and other
issues caused by excessive navigation attempts. The limiter is enforced at innerDispatchNavigateEvent()
to cover all navigation methods (navigate, reload, back, forward, traverseTo).

Tests: navigation-api/navigation-api-rate-limit-exceeded.html
       navigation-api/navigation-api-rate-limit-window-reset.html

* LayoutTests/navigation-api/navigation-api-rate-limit-exceeded-expected.txt: Added.
* LayoutTests/navigation-api/navigation-api-rate-limit-exceeded.html: Added.
* LayoutTests/navigation-api/navigation-api-rate-limit-window-reset-expected.txt: Added.
* LayoutTests/navigation-api/navigation-api-rate-limit-window-reset.html: Added.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::RateLimiter::navigationAllowed):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setNavigationRateLimiterParameters):
(WebCore::Internals::resetNavigationRateLimiter):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/302943@main">https://commits.webkit.org/302943@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38fbf77d16764ffe66e05d28bea7809355a72be3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130678 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/3047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41633 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138102 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82307 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e80b8c82-7886-4408-95dc-dbe803369911) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132549 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2965 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2842 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99560 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/50466ad3-eb55-4e0b-bba5-be907096d936) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133625 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2146 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117011 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80268 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b98a148e-4330-4083-a851-8771cadfc0bb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2070 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35145 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81355 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110649 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140579 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2739 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2475 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108065 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2783 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113354 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107998 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27483 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2110 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31781 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55734 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2809 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66198 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2629 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2830 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2735 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

basuke commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -7295,6 +7296,24 @@ String Internals::windowLocationHost(DOMWindow& window)
 
+void Internals::setNavigationRateLimiterParameters(DOMWindow& window, unsigned maxNavigations, double windowDurationSeconds)
+{
+    RefPtr localWindow = dynamicDowncast<LocalDOMWindow>(window);
```
You should simply write:
```
if (RefPtr localWindow = dynamicDowncast<LocalDOMWindow>(window))
    localWindow->navigation().rateLimiterForTesting(...)
```
to simplify the code.

basuke commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -7295,6 +7296,24 @@ String Internals::windowLocationHost(DOMWindow& window)
+
+void Internals::resetNavigationRateLimiter(DOMWindow& window)
+{
+    RefPtr localWindow = dynamicDowncast<LocalDOMWindow>(window);
```
Ditto

basuke commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -980,6 +980,26 @@ Navigation::DispatchResult Navigation::innerDispatchNavigateEvent(NavigationNavi
+        Ref { tracker->finishedPromise }->reject(exception);
+        cleanupAPIMethodTracker(tracker.get());
+
+        return DispatchResult::Completed;
```
Should we return `DispatchResult::Aborted`?

basuke commented on Source/WebCore/page/Navigation.h:
```diff
@@ -186,6 +188,51 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
+        void markReported() { m_limitMessageSent = true; }
+
+        // Testing support
+        void setParameters(unsigned maxNavigations, Seconds duration)
```
Name it like `setParametersForTesting`.

basuke commented on Source/WebCore/page/Navigation.h:
```diff
@@ -186,6 +188,51 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
+            reset();
+        }
+
+        void reset()
```
Ditto.

basuke reviewed (approved):
LGTM. Thanks!

roberto-apple committed:
Implement navigation rate limiting
https://bugs.webkit.org/show_bug.cgi?id=302355
rdar://164510890

Reviewed by Basuke Suzuki.

Add a sliding window rate limiter (200 navigations per 10 second window) to the Navigation API
to prevent malicious or buggy scripts from flooding the system with navigation requests. This
matches Chromium's implementation and prevents IPC flooding, resource exhaustion, and other
issues caused by excessive navigation attempts. The limiter is enforced at innerDispatchNavigateEvent()
to cover all navigation methods (navigate, reload, back, forward, traverseTo).

Tests: navigation-api/navigation-api-rate-limit-exceeded.html
       navigation-api/navigation-api-rate-limit-window-reset.html

* LayoutTests/navigation-api/navigation-api-rate-limit-exceeded-expected.txt: Added.
* LayoutTests/navigation-api/navigation-api-rate-limit-exceeded.html: Added.
* LayoutTests/navigation-api/navigation-api-rate-limit-window-reset-expected.txt: Added.
* LayoutTests/navigation-api/navigation-api-rate-limit-window-reset.html: Added.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::RateLimiter::navigationAllowed):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setNavigationRateLimiterParameters):
(WebCore::Internals::resetNavigationRateLimiter):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/302943@main
